#Trabalho prático # 4
##Introdução à Computação Numérica

###Tópico: Integração numérica

---

Nome: *Eduarda Leal*

DRE: *118199369*


---

**Instruções:**

**Preencha o notebook** com seus dados, seus códigos e **rode todos os códigos**

**Inclua  um relatório, na parte final do notebook,** com suas conclusões, respostas, observações e comentários relacionados com os tópicos e questões abordadas nas tarefas.

**A elaboração dos códigos pode ser realizada em duplas, mas o resto do trabalho é individual.**

*Se você fez o seu trabalho em dupla inclua essa informação no relatório. Além disso, se você  também recebeu ajuda de um(a) colega/amigo(a) ou consultou/usou informações da internet ou outras fontes, coloque referências à elas no seu relatório.*


**Entrega do trabalho:**

*Data de entrega: 27/junho/2023*

---


## Tarefa #1

Implemente funções em Python para aproximar a integral definida
$$
I = \int_a^b f(x)\,dx,
$$
usando:
i) a regra dos trapézios repetida/composta;
ii) a regra de Simpson repetida/composta.

a) Considere como parâmetros de entrada o espaçamento $h$ entre os nós (pontos) que serão usados e uma lista com os valores da função nesses nós: $f_j = f(a+j\, h)$,  $j=0,\dots,n$.

b) Considere como parâmetros de entrada o intervalo $[a, b]$, a função $f$ e a quantidade de nós (pontos) que serão usados.

**Observação**: Você precisará implementar 4 funções.


In [ ]:
# Função que calcula a integral utilizando o método do trapézio repetido
def trapezio_repetida(h, f):
    n = len(f) - 1
    integral = f[0] + f[n]  # Soma o primeiro e o último valor da função

    for i in range(1, n):  # Percorre os valores intermediários
        integral += 2 * f[i]  # Multiplica por 2 os valores intermediários

    integral *= h / 2  # Multiplica pelo passo da integração dividido por 2
    return integral

# Função que calcula a integral utilizando o método de Simpson repetido
def simpson_repetida(h, f):
    n = len(f) - 1
    integral = f[0] + f[n]  # Soma o primeiro e o último valor da função

    for i in range(1, n):  # Percorre os valores intermediários
        if i % 2 == 0:  # Verifica se o índice é par
            integral += 2 * f[i]  # Multiplica por 2 os valores de índice par
        else:
            integral += 4 * f[i]  # Multiplica por 4 os valores de índice ímpar

    integral *= h / 3  # Multiplica pelo passo da integração dividido por 3
    return integral

# Função que calcula a integral utilizando o método do trapézio composta
def trapezio_composta(a, b, f, n):
    h = (b - a) / n  # Calcula o tamanho do passo da integração
    x = [a + i * h for i in range(n + 1)]  # Cria uma lista de pontos equidistantes
    y = [f(xi) for xi in x]  # Calcula os valores da função nos pontos
    integral = trapezio_repetida(h, y)  # Chama a função trapezio_repetida com os valores calculados
    return integral

# Função que calcula a integral utilizando o método de Simpson composta
def simpson_composta(a, b, f, n):
    h = (b - a) / n  # Calcula o tamanho do passo da integração
    x = [a + i * h for i in range(n + 1)]  # Cria uma lista de pontos equidistantes
    y = [f(xi) for xi in x]  # Calcula os valores da função nos pontos
    integral = simpson_repetida(h, y)  # Chama a função simpson_repetida com os valores calculados
    return integral


##Tarefa #2: Teste as funções implementadas

a) Procure uma função $f$ que não seja um polinômio e cuja integral definida pode ser calculada analiticamente (de forma exata) no intervalo escolhido.

b) Obtenha aproximações do valor da integral definida dessa função considerando $n=10$, $50$, $100$ e $200$ (note que a quantidade de nós em cada caso será igual a $n+1$). Para cada valor de $n$ obtenha aproximações usando suas implementações das regras dos trapézios e de Simpson.

c) Determine os erros absolutos associados com as aproximações obtidas no item b).

d) Comente e explique os resultados obtidos em b) e c).

In [ ]:
import math

f = lambda x: math.sin(x) # Função f(x) = sen(x)

print("Letra A\n")
print("A função escolhida foi f(x) = sen(x)")

# Letra B e C

n_values = [10, 50, 100, 200]
integal_exata = 2

# Listas para armazenar os resultados
aproximacoes = []
erros_absolutos = []

for n in n_values:
    # Intervalo [0, pi]
    a = 0
    b = math.pi

    h = (b - a) / n # Calculando h baseado em a, b e n

    # Criando a lista de nós
    lista = []
    for i in range(n + 1):
        lista.append(f(a + i * h))

    # Aproximando os valores da integral
    aprox_rep_trapezio = trapezio_repetida(h, lista)
    aprox_rep_simpson = simpson_repetida(h, lista)
    aprox_comp_trapezio = trapezio_composta(a, b, f, n)
    aprox_comp_simpson = simpson_composta(a, b, f, n)

    # Armazenando os resultados nas listas
    aproximacoes.append((aprox_rep_trapezio, aprox_rep_simpson, aprox_comp_trapezio, aprox_comp_simpson))

    # Calculando os erros absolutos
    erro_rep_trapezio = abs(aprox_rep_trapezio - integal_exata)
    erro_rep_simpson = abs(aprox_rep_simpson - integal_exata)
    erro_comp_trapezio = abs(aprox_comp_trapezio - integal_exata)
    erro_comp_simpson = abs(aprox_comp_simpson - integal_exata)

    # Armazenando os erros absolutos nas listas
    erros_absolutos.append((erro_rep_trapezio, erro_rep_simpson, erro_comp_trapezio, erro_comp_simpson))


# Imprimindo as aproximações para todos os valores de n
print("\nLetra B\n")
for i, n in enumerate(n_values):
    aprox = aproximacoes[i]
    print(f"For n = {n}:")
    print(f"Aproximação repetida trapezio: {aprox[0]:.5f}")
    print(f"Aproximação composta trapezio: {aprox[2]:.5f}")
    print(f"Aproximação repetida simpson: {aprox[1]:.5f}")
    print(f"Aproximação composta simpson: {aprox[3]:.5f}")
    print("-----------------------------------------")

# Imprimindo os erros absolutos para todos os valores de n
print("\nLetra C\n")
for i, n in enumerate(n_values):
    erros = erros_absolutos[i]
    print(f"For n = {n}:")
    print(f"Erro absoluto da aproximação repetida trapezio: {erros[0]:.5f}")
    print(f"Erro absoluto da aproximação composta trapezio: {erros[2]:.5f}")
    print(f"Erro absoluto da aproximação repetida simpson: {erros[1]:.5f}")
    print(f"Erro absoluto da aproximação composta simpson: {erros[3]:.5f}")
    print("---------------------------------------------------------")


Letra A

A função escolhida foi f(x) = sen(x)

Letra B

For n = 10:
Aproximação repetida trapezio: 1.98352
Aproximação composta trapezio: 1.98352
Aproximação repetida simpson: 2.00011
Aproximação composta simpson: 2.00011
-----------------------------------------
For n = 50:
Aproximação repetida trapezio: 1.99934
Aproximação composta trapezio: 1.99934
Aproximação repetida simpson: 2.00000
Aproximação composta simpson: 2.00000
-----------------------------------------
For n = 100:
Aproximação repetida trapezio: 1.99984
Aproximação composta trapezio: 1.99984
Aproximação repetida simpson: 2.00000
Aproximação composta simpson: 2.00000
-----------------------------------------
For n = 200:
Aproximação repetida trapezio: 1.99996
Aproximação composta trapezio: 1.99996
Aproximação repetida simpson: 2.00000
Aproximação composta simpson: 2.00000
-----------------------------------------

Letra C

For n = 10:
Erro absoluto da aproximação repetida trapezio: 0.01648
Erro absoluto da aproximação com

##Tarefa #3: Aplicação

A **integral de Fresnel C(x)** é uma função definida como
$$
\mathrm{C}(x) = \int_0^x \cos(\pi\,t^2/2)\,dt.
$$

a) Faça a implementação de uma função que determine o valor aproximado de $\mathrm{C}(x)$ para qualquer entrada $x$ com um erro absoluto menor que $10^{-8}$. Explique as ideias básicas da sua implementação no seu relatório.

b) Teste a função implementada comparando com os seguintes valores*:

i) $\mathrm{C}(-2) = -0.48825340607534075450$,

ii) $\mathrm{C}(0.5) = 0.49234422587144639288$,

iii) $\mathrm{C}(1) = 0.77989340037682282947$,

iv) $\mathrm{C}(5) = 0.56363118870401223110$.

*Os valores apresentados foram calculados com 16 dígitos de precisão.




In [ ]:
# LETRA A

import math

n = 2000 #número de subintervalos
f = lambda t: math.cos((math.pi*t**2)/2) # Função f(x) = cos((pi*t**2)/2

def fresnel (x):
  return simpson_composta(0, x, f, n)

#Letra B
print("Testando a função\n")
valores_x = [-2, 0.5, 1, 5]
valores_esperados = [-0.48825340607534075450, 0.49234422587144639288, 0.77989340037682282947, 0.56363118870401223110]

for i, x in enumerate(valores_x):
    aprox_fresnel = fresnel(x)
    valor_esperado = valores_esperados[i]
    erro_absoluto = abs(aprox_fresnel - valor_esperado)
    print(f"C({x}) = {aprox_fresnel}, Erro absoluto = {erro_absoluto:.10f}")


Testando a função

C(-2) = -0.4882534060750135, Erro absoluto = 0.0000000000
C(0.5) = 0.49234422587144505, Erro absoluto = 0.0000000000
C(1) = 0.7798934003768333, Erro absoluto = 0.0000000000
C(5) = 0.5636311895452666, Erro absoluto = 0.0000000008


## Relatório

### **Tarefa 1**

Foi feita a implementação dos métodos do trapezio e Simpson cada um na forma repetida e composta. Os códigos se encontram feitos e comentados na primeira seção do trabalho.

### **Tarefa 2**
### Letra A

A função escolhida foi f(x) = sen(x) no intervalo de [0, pi] e sua integral definida é 2.

### Letra B e C

Para n = 10, 50, 100 e 200, os resultados da aproximação (calculados na letra b) e seus respectivos erros absolutos (calculados na letra c), são:

**Para n = 10:**

Aproximação repetida trapezio: 1.98352 Erro absoluto:0.01648

Aproximação composta trapezio: 1.98352 Erro absoluto:0.01648

Aproximação repetida simpson: 2.00011 Erro absoluto: 0.00011

Aproximação composta simpson: 2.00011 Erro absoluto: 0.00011


**Para n = 50:**

Aproximação repetida trapezio: 1.99934 Erro absoluto:0.00066

Aproximação composta trapezio: 1.99934 Erro absoluto:0.00066

Aproximação repetida simpson: 2.00000 Erro absoluto: 0.00000

Aproximação composta simpson: 2.00000 Erro absoluto: 0.00000


**Para n = 100:**

Aproximação repetida trapezio: 1.99984 Erro absoluto:0.00016

Aproximação composta trapezio: 1.99984 Erro absoluto:0.00016

Aproximação repetida simpson: 2.00000 Erro absoluto: 0.00000

Aproximação composta simpson: 2.00000 Erro absoluto: 0.00000


**Para n = 200:**

Aproximação repetida trapezio: 1.99996 Erro absoluto:0.00004

Aproximação composta trapezio: 1.99996 Erro absoluto:0.00004

Aproximação repetida simpson: 2.00000 Erro absoluto: 0.00000

Aproximação composta simpson: 2.00000 Erro absoluto: 0.00000



### Letra D

Ao analisar os resultados obtidos, é percepitivel que ao aumentar o número de subintervalos, os erros absolutos ficam extremamente pequenos, resultando em aproximações altamente precisas. E isso é válido tanto para a regra do trapézio quanto a regra de Simpson, porém, a regra de Simpson fornece aproximações mais precisas do que a regra dos trapézios com o mesmo número de subintervalos.

### **Tarefa 3**

###Letra A

Para fazer a função fresnel, resolvi usar o método da regra de Simpson composta com n = 2000, pois foi o método que obteve melhores resultados na tarefa 2.

###Letra B

Ao realizar os teste, consegui encontrar os valores bem proximo do esperado e com os erros absolutos menores que $10^{-8}$
Os valores encontrados e os erros absolutos foram:

C(-2) = -0.4882534060753374, Erro absoluto = 0.00000000

C(0.5) = 0.4923442258714486, Erro absoluto = 0.00000000

C(1) = 0.77989340037683, Erro absoluto = 0.00000000

C(5) = 0.5636311887040968, Erro absoluto = 0.00000000